src/Framework/Constraint/Count.php

```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
             $this->expectedCount
         );
     }
+
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->traversableCounts === null) {
+            $this->traversableCounts = new \SplObjectStorage();
+        }
+
+        while ($traversable instanceof IteratorAggregate) {
+            $traversable = $traversable->getIterator();
+        }
+
+        if ($this->traversableCounts->contains($traversable)) {
+            return $this->traversableCounts[$traversable];
+        }
+
+        if ($traversable instanceof Iterator) {
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
+        }
+
+        return $this->getCountOfTraversableThatIsNotIteratorOrIteratorAggregate($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key   = $iterator->key();
+        $count = $this->getCountOfIterator($iterator);
+
+        $this->traversableCounts->attach($iterator, $count);
+
+        if ($this->iteratorIsRewindable($iterator)) {
+            $this->traversableCounts->detach($iterator);
+            $this->rewindIterator($iterator, $key);
+        }
+
+        return $count;
+    }
+
+    private function iteratorIsRewindable(Iterator $iterator): bool
+    {
+        try {
+            $iterator->rewind();
+        } catch (\Exception $e) {
+            return false;
+        }
+
+        return !($iterator instanceof \NoRewindIterator);
+    }
+
+    /**
+     * Returns the total number of iterations from a iterator.
+     * This will fully exhaust the iterator.
+     */
+    private function getCountOfIterator(Iterator $iterator): int
+    {
+        for ($countOfGenerator = 0; $iterator->valid(); $iterator->next()) {
+            ++$countOfGenerator;
+        }
+
+        return $countOfGenerator;
+    }
+
+    private function rewindIterator(Iterator $iterator, $key): void
+    {
+        $iterator->rewind();
+
+        while ($iterator->valid() && $key !== $iterator->key()) {
+            $iterator->next();
+        }
+    }
+
+    private function getCountOfTraversableThatIsNotIteratorOrIteratorAggregate(Traversable $traversable): int
+    {
+        $count = \iterator_count($traversable);
+
+        $this->traversableCounts->attach($traversable, $count);
```

Comment:
Maybe it is possible to factorize this `attach` with the other (i.e. directly in `function getCountOfTraversable`)

Reply:
I think that would make other things less nice and not be a good trade-off